gcloud-java-storage/src/main/java/com/google/gcloud/storage/BlobWriteChannelImpl.java

```diff
@@ -150,4 +154,120 @@ public void chunkSize(int chunkSize) {
     chunkSize = (chunkSize / MIN_CHUNK_SIZE) * MIN_CHUNK_SIZE;
     this.chunkSize = Math.max(MIN_CHUNK_SIZE, chunkSize);
   }
+
+  static class StateImpl implements RestorableState<BlobWriteChannel>, Serializable {
+
+    private static final long serialVersionUID = 8541062465055125619L;
+
+    private final StorageOptions serviceOptions;
+    private final BlobInfo blobInfo;
+    private final String uploadId;
+    private final int position;
+    private final byte[] buffer;
+    private final int limit;
+    private final boolean isOpen;
+    private final int chunkSize;
+
+    StateImpl(Builder builder) {
+      this.serviceOptions = builder.serviceOptions;
+      this.blobInfo = builder.blobInfo;
+      this.uploadId = builder.uploadId;
+      this.position = builder.position;
+      this.buffer = builder.buffer;
+      this.limit = this.buffer.length;
+      this.isOpen = builder.isOpen;
+      this.chunkSize = builder.chunkSize;
+    }
+
+    static class Builder {
+      private final StorageOptions serviceOptions;
+      private final BlobInfo blobInfo;
+      private final String uploadId;
+      private int position;
+      private byte[] buffer;
+      private int limit;
+      private boolean isOpen;
+      private int chunkSize;
+
+      private Builder(StorageOptions options, BlobInfo blobInfo, String uploadId) {
+        this.serviceOptions = options;
+        this.blobInfo = blobInfo;
+        this.uploadId = uploadId;
+      }
+
+      Builder position(int position) {
+        this.position = position;
+        return this;
+      }
+
+      Builder buffer(byte[] buffer) {
+        this.buffer = buffer;
+        return this;
+      }
+
+      Builder isOpen(boolean isOpen) {
+        this.isOpen = isOpen;
+        return this;
+      }
+
+      Builder chunkSize(int chunkSize) {
+        this.chunkSize = chunkSize;
+        return this;
+      }
+
+      RestorableState<BlobWriteChannel> build() {
+        return new StateImpl(this);
+      }
+    }
+
+    static Builder builder(StorageOptions options, BlobInfo blobInfo, String uploadId) {
+      return new Builder(options, blobInfo, uploadId);
+    }
+
+    @Override
+    public BlobWriteChannel restore() {
+      BlobWriteChannelImpl channel = new BlobWriteChannelImpl(serviceOptions, blobInfo, uploadId);
+      channel.position = position;
+      channel.buffer = buffer.clone();
+      channel.limit = limit;
```

Comment:
limit is unnecessary and `channel.limit` could be assign with `buffer.capacity()`